geps/gep-1494/index.md

```diff
@@ -130,20 +130,383 @@ From @ongy, some additional goals to keep in mind:
+
+	// HTTPAuthConfig contains configuration for communication with HTTP-speaking
+	// backends.
+	//
+	// If unset, implementations must assume the default behavior for each
+	// included field is intended.
+	//
+	// +optional
+	HTTPAuthConfig *HTTPAuthConfig `json:"http,omitempty"`
+
+	// ForwardBody controls if requests to the authorization server should include
+	// the body of the client request; and if so, how big that body is allowed
+	// to be.
+	//
+	// It is expected that implementations will buffer the request body up to
+	// `forwardBody.maxSize` bytes. Bodies over that size must be rejected with a
+	// 4xx series error (413 or 403 are common examples), and fail processing
+	// of the filter.
+	//
+	// If unset, or `forwardBody.maxSize` is set to `0`, then the body will not
+	// be forwarded.
+	//
+	// Feature Name: HTTPRouteExtAuthForwardBody
+	//
+	// GEP Review Notes:
+	// Both Envoy and Traefik show support for this feature, but HAProxy and
+	// ingress-nginx do not. So this has a separate feature flag for it.
+	//
+	// +optional
+	ForwardBody *ForwardBodyConfig `json:"forwardBody,omitempty"`
+}
+
+// GRPCAuthConfig contains configuration for communication with ext_auth
+// protocol-speaking backends.
+type GRPCAuthConfig struct {
+
+	// AllowedRequestHeaders specifies what headers from the client request
+	// will be sent to the authorization server.
+	//
+	// If this list is empty, then all headers must be sent.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxLength=64
+	AllowedRequestHeaders []string `json:"allowedHeaders,omitempty"`
+}
+
+// HTTPAuthConfig contains configuration for communication with HTTP-speaking
+// backends.
+type HTTPAuthConfig struct {
+	// Path sets the prefix that paths from the client request will have added
+	// when forwarrded to the authorization server.
+	//
+	// When empty or unspecified, no prefix is added.
+	// +optional
+	Path string `json:"path,omitempty"`
+
+	// AllowedRequestHeaders specifies what additional headers from the client request
+	// will be sent to the authorization server.
+	//
+	// The following headers must always be sent to the authorization server,
+	// regardless of this setting:
+	// 
+	// * `Host`
+	// * `Method`
+	// * `Path`
+	// * `Content-Length`
+	// * `Authorization`
+	//
+	// If this list is empty, then only those headers must be sent.
+	//
+	// Nick's GEP Review Notes:
+	// Envoy actually does not allow sending all headers to HTTP authorization
+	// servers. If the equivalent Envoy field is unspecified, then only
+	// `Host`, `Method`, `Path`, `Content-Length` and `Authorization` are included.
+	// This is kept for backwards compatibility in Envoy.
+	//
+	// Traefik includes _all_ headers when the equivalent setting is empty.
+	//
+	// This definition is the only compromise I can see between the two; Envoy
+	// can set this to empty, and Traefik can always explicitly specify those headers
+	// at a minimum.
```

Comment:
🤔 Pretty sure that both Envoy and Traefik can just as easily set the field explicitly?

Back in Emissaryland we used to explicitly always include 

- `Authorization`
- `Cookie`
- `From`
- `Proxy-Authorization`
- `User-Agent`
- `X-Forwarded-For`
- `X-Forwarded-Host`
- `X-Forwarded-Proto`

and of course `Content-Length` as described above, for the record.

Reply:
Yes, I based this on the default config for when the corresponding field is left empty in each implementation. But we can just mandate that, when _this_ field is left empty, it must be translated to a specific list, and have that work across implementations.

Reply:
I've removed the GEP Review comments, and left the headers sent for "" to the original set.

If you really feel we should change that default, please feel free to comment and I'll do so.